transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.channel;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.util.internal.ObjectUtil;
+
+import java.util.ArrayDeque;
+
+/**
+ * A utility class which allows a buffer producer to enqueue arbitrarily sized buffers with promises
+ * and the consumer to dequeue differently sized buffers with the guarantee that when all the bytes in an added
+ * buffer are removed the promise associated with it will be completed when promise passed in during removal of the
+ * last byte completes.
+ *
+ * <p>This functionality is useful for aggregating or partitioning writes into fixed size buffers for framing protocols
+ * such as HTTP2.
+ */
+public class CoalescingBufferQueue {
+
+    private final ByteBufAllocator alloc;
+    private final ArrayDeque<Object> promisesAndBufs = new ArrayDeque<Object>();
```

Comment:
Hard to know what a good default is, perhaps allow an alternate constructor that takes initial size?
Alternatively use Recyclable linked-list entries (see other comment)